crates/core/tedge_actors/src/test_helpers.rs

```diff
@@ -7,130 +7,307 @@ use crate::MessageSource;
 use crate::NoConfig;
 use crate::NullSender;
 use crate::ReceiveMessages;
+use crate::RuntimeRequest;
 use crate::Sender;
 use crate::ServiceConsumer;
 use crate::ServiceProvider;
 use crate::SimpleMessageBox;
 use crate::SimpleMessageBoxBuilder;
+use crate::WrappedInput;
+use async_trait::async_trait;
 use futures::stream::FusedStream;
 use futures::SinkExt;
 use futures::StreamExt;
-use std::collections::HashSet;
 use std::fmt::Debug;
-use std::hash::Hash;
 use std::time::Duration;
 
-/// Check that all messages are received in the given order without any interleaved messages.
-///
-/// Returns early on `timeout` while waiting for the next message.
-///
-/// ```rust
-/// use crate::tedge_actors::{Builder, NoMessage, RuntimeError, ServiceConsumer, SimpleMessageBox, SimpleMessageBoxBuilder, test_helpers};
-///
-/// #[derive(Debug,Eq,PartialEq,Hash)]
-/// enum MyMessage {
-///    Foo(u32),
-///    Bar(u32),
-/// }
-///
-/// # #[tokio::main]
-/// # async fn main() -> Result<(),RuntimeError> {
-///
-/// use std::time::Duration;
-/// let mut receiver_builder = SimpleMessageBoxBuilder::new("Recv", 16);
-/// let sender_builder = SimpleMessageBoxBuilder::new("Send", 16).with_connection(&mut receiver_builder);
-/// let mut sender = sender_builder.build();
-/// let mut receiver: SimpleMessageBox<MyMessage,NoMessage> = receiver_builder.build();
-///
-/// sender.send(MyMessage::Foo(1)).await?;
-/// sender.send(MyMessage::Bar(2)).await?;
-/// sender.send(MyMessage::Foo(3)).await?;
-///
-/// test_helpers::assert_received(&mut receiver, Duration::from_millis(100), [
-///     MyMessage::Foo(1),
-///     MyMessage::Bar(2),
-///     MyMessage::Foo(3),
-/// ]);
-///
-/// # Ok(())
-/// # }
-///
-/// ```
-pub async fn assert_received<T, U>(
-    messages: &mut impl ReceiveMessages<U>,
-    timeout: Duration,
-    expected: T,
-) where
-    T: IntoIterator,
-    U: From<T::Item>,
-    U: Debug + Eq,
+/// A test helper that extends a message box with various way to check received messages.
+#[async_trait]
+pub trait MessageReceiverExt<M: Message> {
+    /// Return a new receiver which returns None if no message is received after the given timeout
```

Comment:
The issue is that the returned receiver must implement  ~ReceiveMessages~  `MessageReceiver`.

In practice, I don't see that as an issue for testing. Either the sender has been closed prematurely  while the test is expecting a value or the sender failed to send a message without the given time window. In both case, the test received nothing aka `None`.

This would matter if not used for testing purpose though. The ~ReceiveMessages~  `MessageReceiver` trait can be extended with `recv` method with a timeout.